core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala

```diff
@@ -281,6 +296,202 @@ class HistoryServerSuite extends SparkFunSuite with BeforeAndAfter with Matchers
+
+      // which lists as incomplete
+      assertUrlContains(historyServerIncompleted, appId)
+
+      val appIdRoot = buildURL(appId, "")
+      val rootAppPage = HistoryServerSuite.getUrl(appIdRoot)
+      logDebug(s"$appIdRoot ->[${rootAppPage.length}] \n$rootAppPage")
+      // sanity check to make sure filter is chaining calls
+      rootAppPage should not be empty
+
+      def getAppUI: SparkUI = {
+        provider.getAppUI(appId, None).get.ui
+      }
+
+      // selenium isn't that useful on failures...add our own reporting
+      def getNumJobs(suffix: String): Int = {
+        val target = buildURL(appId, suffix)
+        val targetBody = HistoryServerSuite.getUrl(target)
+        try {
+          go to target.toExternalForm
+          findAll(cssSelector("tbody tr")).toIndexedSeq.size
+        } catch {
+          case ex: Exception =>
+            throw new Exception(s"Against $target\n$targetBody", ex)
+        }
+      }
+      def completedJobs(): Seq[JobUIData] = {
+        getAppUI.jobProgressListener.completedJobs
+      }
+      def activeJobs(): Seq[JobUIData] = {
+        getAppUI.jobProgressListener.activeJobs.values.toSeq
+      }
+
+      activeJobs() should have size 0
+      completedJobs() should have size 1
+      getNumJobs("") should be(1)
+      getNumJobs("/jobs") should be(1)
+      assert(metrics.lookupCount.getCount > 1, s"lookup count too low in $metrics")
+
+      // dump state before the next bit of test, which is where update
+      // checking really gets stressed
+      dumpLogDir("filesystem before executing second job")
+      logDebug(s"History Server: $server")
+
+      // second job needs a delay to ensure timestamps are different
+      Thread.sleep(10)
+
+      logDebug("Starting second job")
+      val d2 = sc.parallelize(1 to 10)
+      d2.count()
+      dumpLogDir("After second job")
+
+
+      val stdTimeout = timeout(10 seconds)
+
+      logDebug("waiting for UI to update")
+      eventually(stdTimeout, stdInterval) {
+        assert(2 === getNumJobs(""),
+          s"jobs not updated, server=$server\n dir = ${listDir(logDirPath)}")
+        assert(2 === getNumJobs("/jobs"),
+          s"job count under /jobs not updated, server=$server\n dir = ${listDir(logDirPath)}")
+        }
+
+      // and again, without any intermediate sleep, so relying on size changes rather than modtime
+      d.count()
+      d.count()
+      eventually(stdTimeout, stdInterval) {
+        assert(4 === getNumJobs("/jobs"),
+          s"two jobs back-to-back not updated, server=$server\n")
+      }
+      val startcount = getNumJobs("/jobs")
+      // do a series with different sleep times to see if that can create trouble
+      val limit = 10
+      for (i <- 1 to limit) {
+        d.count()
+        Thread.sleep(i * 100)
+        eventually(stdTimeout, stdInterval) {
+          assert(startcount + i === getNumJobs("/jobs"),
+            s"jobs iteration $i not updated, server=$server\n")
+        }
+      }
```

Comment:
I  know they are probably redundant, I just wanted to make sure that back-to-back calls weren't hiding anything. I can cut them